addons/binding/org.openhab.binding.regoheatpump/src/main/java/org/openhab/binding/regoheatpump/handler/Rego6xxHeatPumpHandler.java

```diff
@@ -0,0 +1,405 @@
+import org.eclipse.smarthome.core.types.RefreshType;
+import org.eclipse.smarthome.core.types.State;
+import org.eclipse.smarthome.core.types.UnDefType;
+import org.openhab.binding.regoheatpump.internal.protocol.RegoConnection;
+import org.openhab.binding.regoheatpump.internal.rego6xx.CommandFactory;
+import org.openhab.binding.regoheatpump.internal.rego6xx.ErrorLine;
+import org.openhab.binding.regoheatpump.internal.rego6xx.RegoRegisterMapper;
+import org.openhab.binding.regoheatpump.internal.rego6xx.ResponseParser;
+import org.openhab.binding.regoheatpump.internal.rego6xx.ResponseParserFactory;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link Rego6xxHeatPumpHandler} is responsible for handling commands, which are
+ * sent to one of the channels.
+ *
+ * @author Boris Krivonog - Initial contribution
+ */
+abstract class Rego6xxHeatPumpHandler extends BaseThingHandler {
+
+    private static final class ChannelDescriptor {
+        private Date lastUpdate;
+        private byte[] cachedValue;
+
+        public byte[] cachedValueIfNotExpired(int refreshTime) {
+            if (lastUpdate == null || (lastUpdate.getTime() + refreshTime * 900 < new Date().getTime())) {
+                return null;
+            }
+
+            return cachedValue;
+        }
+
+        public void setValue(byte[] value) {
+            lastUpdate = new Date();
+            cachedValue = value;
+        }
+    }
+
+    private final Logger logger = LoggerFactory.getLogger(Rego6xxHeatPumpHandler.class);
+    private final Map<String, ChannelDescriptor> channelDescriptors = new HashMap<>();
+    private int refreshInterval;
+    private RegoConnection connection;
+    private RegoRegisterMapper mapper;
+    private ScheduledFuture<?> scheduledRefreshFuture;
+
+    protected Rego6xxHeatPumpHandler(Thing thing) {
+        super(thing);
+    }
+
+    protected abstract RegoConnection createConnection();
+
+    @Override
+    public void initialize() {
+        mapper = RegoRegisterMapper.rego600();
+        refreshInterval = ((Number) getConfig().get(REFRESH_INTERVAL)).intValue();
+
+        connection = createConnection();
+        logger.info("Rego controller connected via {}", connection.connectionInfo());
+
+        scheduledRefreshFuture = scheduler.scheduleWithFixedDelay(this::refresh, 2, refreshInterval, TimeUnit.SECONDS);
+
+        updateStatus(ThingStatus.UNKNOWN);
+    }
+
+    @Override
+    public void dispose() {
+        super.dispose();
+
+        connection.close();
+
+        scheduledRefreshFuture.cancel(true);
+        scheduledRefreshFuture = null;
+
+        synchronized (channelDescriptors) {
+            channelDescriptors.clear();
+        }
+
+        connection = null;
+        mapper = null;
+    }
+
+    @Override
+    public void handleCommand(ChannelUID channelUID, Command command) {
+        if (command instanceof RefreshType) {
+
+            scheduledRefreshFuture.cancel(false);
```

Comment:
Please note that each read is blocking and takes ~500ms. AFAIK handleCommand is called in a for loop for each linked channel (at least i.e. on startup). So if one has ~10+ linked channels, it will take 5+ seconds in order to exit the for loop (that invoked handleCommand for each linked channel) -> resulting in the "Dispatching event to subscriber 'org.eclipse.smarthome.core.thing.link.ThingLinkManager@1ae8dd3' takes more than 5000ms.".

So this re-scheduling is here to invoke the refresh ASAP and process it in a separate thread so the above described behavior does not occur.

I replaced it with a direct call in the latest code change, as suggested, but not sure if this is the best way to do it, as explained. There is nothing wrong with as it is now in the latest version (functional wise), the downside is that it blocks the thread that invoked the handleCommand. Please advise.